hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala

```diff
@@ -0,0 +1,61 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.virtual.{TArray, TIterable, TStream}
+
+object LowerArrayToStream {
+  private def nonStreamIterable(node: IR) = node.typ.isInstanceOf[TIterable] && !node.typ.isInstanceOf[TStream]
+
+  private def boundary(node: IR): IR = {
+    var streamified = streamify(node)
+
+    if (streamified.typ.isInstanceOf[TStream] && node.typ.isInstanceOf[TArray])
+      streamified = ToArray(streamified)
+
+    assert(streamified.typ == node.typ)
+    streamified
+  }
+
+  private def toStream(node: IR): IR = {
+    val r = streamify(node)
+    if(nonStreamIterable(r))
+      ToStream(r)
+    else
+      r
+  }
+
+  private def streamify(node: IR): IR = {
+    node match {
+      case ArrayFold(a, zero, zn, an, body) => ArrayFold(toStream(a), boundary(zero), zn, an, boundary(body))
+      case ArrayFor(a, n, b) => ArrayFor(toStream(a), n, boundary(b))
+      case ArrayFold2(a, accum, valueName, seq, result) => ArrayFold2(
+        toStream(a),
+        accum.map { case (name, value) => (name, boundary(value)) },
+        valueName, seq.map(boundary), boundary(result))
+      case RunAggScan(a, name, init, seq, res, sig) => RunAggScan(toStream(a), name, boundary(init), boundary(seq), boundary(res), sig)
+      case MakeArray(args, t) => MakeStream(args.map(boundary), TStream(t.elementType, t.required))
+      case ArrayRange(start, stop, step) => StreamRange(boundary(start), boundary(stop), boundary(step))
+      case ArrayZip(childIRs, names, body, behavior) => ArrayZip(childIRs.map(toStream), names, boundary(body), behavior)
+      case ArrayMap(a, n, b) => ArrayMap(toStream(a), n, boundary(b))
+      case ArrayFilter(a, n, b) => ArrayFilter(toStream(a), n, boundary(b))
+      case ArrayFlatMap(a, n, b) => ArrayFlatMap(toStream(a), n, toStream(b))
+      case ArrayScan(a, zero, zn, an, body) => ArrayScan(toStream(a), boundary(zero), zn, an, boundary(body))
+      case ArrayLeftJoinDistinct(l, r, ln, rn, keyf, joinf) =>
+        ArrayLeftJoinDistinct(toStream(l), toStream(r), ln, rn, boundary(keyf), boundary(joinf))
+      case x: ApplyIR => boundary(x.explicitNode)
```

Comment:
I kept this in. Without this case (namely, x.explicitNode), testArrayAggContexts breaks. Haven't entirely figured out why I see this, because I don't see where the ArrayFlatMap gets generated (maybe by expansion of ApplyIR "append"?). In any case, ToStream(x.explicitNode) was used in the previous Streamify, so this is consistent.


Before LowerToStream: MakeTuple(ArrayBuffer((0,Let(__iruid_76,ApplyBinaryPrimOp(Multiply(),GetTupleElement(In(0,PCTuple[0:PInt32]),0),GetTupleElement(In(0,PCTuple[0:PInt32]),0)),Let(__iruid_77,RunAgg(Begin(ArrayBuffer(Begin(ArrayBuffer(InitOp(0,ArrayBuffer(),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))), ArrayFor(ArrayRange(I32(0),I32(10),I32(1)),__iruid_78,Let(__iruid_79,Cast(ApplyBinaryPrimOp(Multiply(),ApplyBinaryPrimOp(Multiply(),Ref(__iruid_76,int32),Ref(__iruid_76,int32)),Ref(__iruid_78,int32)),int64),Begin(ArrayBuffer(SeqOp(0,ArrayBuffer(MakeArray(ArrayBuffer(Ref(__iruid_76,int32), Ref(__iruid_78,int32), Cast(Ref(__iruid_79,int64),int32), Cast(Ref(__iruid_79,int64),int32)),array<int32>)),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))))))),ResultOp(0,WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),ApplyIR(append,ArrayBuffer(GetTupleElement(Ref(__iruid_77,tuple(array<array<int32>>)),0), MakeArray(ArrayBuffer(Ref(__iruid_76,int32)),array<int32>))))))))

After LowerArrayToStream MakeTuple(ArrayBuffer((0,Let(__iruid_76,ApplyBinaryPrimOp(Multiply(),GetTupleElement(In(0,PCTuple[0:PInt32]),0),GetTupleElement(In(0,PCTuple[0:PInt32]),0)),Let(__iruid_77,RunAgg(Begin(ArrayBuffer(Begin(ArrayBuffer(InitOp(0,ArrayBuffer(),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))), ArrayFor(StreamRange(I32(0),I32(10),I32(1)),__iruid_78,Let(__iruid_79,Cast(ApplyBinaryPrimOp(Multiply(),ApplyBinaryPrimOp(Multiply(),Ref(__iruid_76,int32),Ref(__iruid_76,int32)),Ref(__iruid_78,int32)),int64),Begin(ArrayBuffer(SeqOp(0,ArrayBuffer(ToArray(MakeStream(ArrayBuffer(Ref(__iruid_76,int32), Ref(__iruid_78,int32), Cast(Ref(__iruid_79,int64),int32), Cast(Ref(__iruid_79,int64),int32)),stream<int32>))),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))))))),ResultOp(0,WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),ApplyIR(append,ArrayBuffer(GetTupleElement(Ref(__iruid_77,tuple(array<array<int32>>)),0), ToArray(MakeStream(ArrayBuffer(Ref(__iruid_76,int32)),stream<int32>)))))))))


Post Lowering and simplifiy: MakeTuple(ArrayBuffer((0,Let(__iruid_118,ApplyBinaryPrimOp(Multiply(),GetTupleElement(In(0,PCTuple[0:PInt32]),0),GetTupleElement(In(0,PCTuple[0:PInt32]),0)),Let(__iruid_119,RunAgg(Begin(ArrayBuffer(Begin(ArrayBuffer(InitOp(0,ArrayBuffer(),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))), ArrayFor(StreamRange(I32(0),I32(10),I32(1)),__iruid_120,Let(__iruid_121,Cast(ApplyBinaryPrimOp(Multiply(),ApplyBinaryPrimOp(Multiply(),Ref(__iruid_118,int32),Ref(__iruid_118,int32)),Ref(__iruid_120,int32)),int64),Begin(ArrayBuffer(SeqOp(0,ArrayBuffer(ToArray(MakeStream(ArrayBuffer(Ref(__iruid_118,int32), Ref(__iruid_120,int32), Cast(Ref(__iruid_121,int64),int32), Cast(Ref(__iruid_121,int64),int32)),stream<int32>))),AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None),Collect()))))))),ResultOp(0,WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),WrappedArray(AggStateSignature(Map(Collect() -> AggSignature(Collect(),ArrayBuffer(),ArrayBuffer(array<int32>))),Collect(),None))),ApplyIR(append,ArrayBuffer(GetTupleElement(Ref(__iruid_119,tuple(array<array<int32>>)),0), ToArray(MakeStream(ArrayBuffer(Ref(__iruid_118,int32)),stream<int32>))))))))) 
2020-02-16 15:24:41 Hail: ERROR: error from strategy JvmCompile



java.lang.ClassCastException: is.hail.expr.types.physical.PCanonicalArray cannot be cast to is.hail.expr.types.physical.PStream

Reply:
you definitely need this in your PR, yeah. 

https://github.com/hail-is/hail/pull/8067 is the medium-term solution, though, rather than unwrapping apply nodes in streamify.